libkbfs/mdserver_disk.go

```diff
@@ -120,445 +89,176 @@ func NewMDServerTempDir(config Config) (*MDServerDisk, error) {
-		return id, nil, errors.New("MD server already shut down")
-	}
-
-	handleBytes, err := md.config.Codec().Encode(handle)
-	if err != nil {
-		return id, nil, err
-	}
+var errMDServerDiskShutdown = errors.New("MDServerDisk is shutdown")
 
-	buf, err := md.handleDb.Get(handleBytes, nil)
-	if err != nil && err != leveldb.ErrNotFound {
-		return id, nil, MDServerError{err}
-	}
-	if err == nil {
-		var id TlfID
-		err := id.UnmarshalBinary(buf)
-		if err != nil {
-			return NullTlfID, nil, err
+func (md *MDServerDisk) getStorage(tlfID TlfID) (*mdServerTlfStorage, error) {
+	storage, err := func() (*mdServerTlfStorage, error) {
+		md.lock.RLock()
+		defer md.lock.RUnlock()
+		if md.tlfStorage == nil {
+			return nil, errMDServerDiskShutdown
 		}
-		rmds, err := md.GetForTLF(ctx, id, NullBranchID, mStatus)
-		return id, rmds, err
-	}
+		return md.tlfStorage[tlfID], nil
+	}()
 
-	// Non-readers shouldn't be able to create the dir.
-	_, uid, err := md.config.KBPKI().GetCurrentUserInfo(ctx)
 	if err != nil {
-		return id, nil, err
-	}
-	if !handle.IsReader(uid) {
-		return id, nil, MDServerErrorUnauthorized{}
-	}
-
-	// Allocate a new random ID.
-	id, err = md.config.Crypto().MakeRandomTlfID(handle.IsPublic())
-	if err != nil {
-		return id, nil, MDServerError{err}
+		return nil, err
 	}
 
-	err = md.handleDb.Put(handleBytes, id.Bytes(), nil)
-	if err != nil {
-		return id, nil, MDServerError{err}
+	if storage != nil {
+		return storage, nil
 	}
-	return id, nil, nil
-}
 
-// GetForTLF implements the MDServer interface for MDServerDisk.
-func (md *MDServerDisk) GetForTLF(ctx context.Context, id TlfID,
-	bid BranchID, mStatus MergeStatus) (*RootMetadataSigned, error) {
-	md.shutdownLock.RLock()
-	defer md.shutdownLock.RUnlock()
-	if *md.shutdown {
-		return nil, errors.New("MD server already shut down")
+	md.lock.Lock()
+	defer md.lock.Unlock()
+	if md.tlfStorage == nil {
+		return nil, errMDServerDiskShutdown
 	}
 
-	if mStatus == Merged && bid != NullBranchID {
-		return nil, MDServerErrorBadRequest{Reason: "Invalid branch ID"}
+	storage = md.tlfStorage[tlfID]
+	if storage != nil {
+		return storage, nil
 	}
 
-	mergedMasterHead, err :=
-		md.getHeadForTLF(ctx, id, NullBranchID, Merged)
+	path := filepath.Join(md.dirPath, tlfID.String())
+	storage, err = makeMDServerTlfStorage(
+		md.codec, md.clock, md.crypto, path)
 	if err != nil {
-		return nil, MDServerError{err}
+		return nil, err
 	}
 
-	// Check permissions
-	ok, err := isReader(
-		ctx, md.config.Codec(), md.config.KBPKI(), mergedMasterHead)
+	md.tlfStorage[tlfID] = storage
+	return storage, nil
+}
+
+func (md *MDServerDisk) getHandleID(ctx context.Context, handle BareTlfHandle,
+	mStatus MergeStatus) (tlfID TlfID, created bool, err error) {
+	handleBytes, err := md.codec.Encode(handle)
 	if err != nil {
-		return nil, MDServerError{err}
+		return NullTlfID, false, MDServerError{err}
 	}
-	if !ok {
-		return nil, MDServerErrorUnauthorized{}
+
+	md.lock.Lock()
+	defer md.lock.Unlock()
```

Comment:
Would it be safe to unlock immediately after the nil check below, so we don't hold the lock while blocking on leveldb access?